<a href="https://colab.research.google.com/github/amd3897/Prediction-of-Product-Sales/blob/main/Prediction_of_product_sales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Predection of Product Sales
- Aziz M Duaibes

## Project overview

### Prediction of product sales
- Potential stakeholders: Grocery stores owners
- Primary goal: Make prediction of sales to increase profit
- What do they need?: Actionable insights/recommendations for which modifications they can make to increase their profits


## Load and Inspect Data

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [1]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [25]:
import pandas as pd
fname = '/content/drive/MyDrive/CodingDojo/01-Fundamentals/Week02/Data/sales_predictions_2023.csv'
df = pd.read_csv(fname)
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [27]:
df = df.set_index('Item_Identifier')
df.head()

Item_Weight Item_Fat_Content  Item_Visibility  \
Item_Identifier                                                  
FDA15                   9.30          Low Fat         0.016047   
DRC01                   5.92          Regular         0.019278   
FDN15                  17.50          Low Fat         0.016760   
FDX07                  19.20          Regular         0.000000   
NCD19                   8.93          Low Fat         0.000000   

                             Item_Type  Item_MRP Outlet_Identifier  \
Item_Identifier                                                      
FDA15                            Dairy  249.8092            OUT049   
DRC01                      Soft Drinks   48.2692            OUT018   
FDN15                             Meat  141.6180            OUT049   
FDX07            Fruits and Vegetables  182.0950            OUT010   
NCD19                        Household   53.8614            OUT013   

                 Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
Item_Identifier                                                               
FDA15                                 1999      Medium               Tier 1   
DRC01                                 2009      Medium               Tier 3   
FDN15                                 1999      Medium               Tier 1   
FDX07                                 1998         NaN               Tier 3   
NCD19                                 1987        High               Tier 3   

                       Outlet_Type  Item_Outlet_Sales  
Item_Identifier                                        
FDA15            Supermarket Type1          3735.1380  
DRC01            Supermarket Type2           443.4228  
FDN15            Supermarket Type1          2097.2700  
FDX07                Grocery Store           732.3800  
NCD19            Supermarket Type1           994.7052

## Clean Data

- There are 8523 rows and 12 columns
- The datatypes are float(4), integer(1), string/object(7)

In [11]:
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [28]:
# check duplicates
df.duplicated().sum()
# no duplicated rows

0

In [29]:
# double check
df[df.duplicated()]

Empty DataFrame
Columns: [Item_Weight, Item_Fat_Content, Item_Visibility, Item_Type, Item_MRP, Outlet_Identifier, Outlet_Establishment_Year, Outlet_Size, Outlet_Location_Type, Outlet_Type, Item_Outlet_Sales]
Index: []

In [31]:
# identufy missing value
df.isna().sum()
# two columns have missing values (Item_Weight(float), Outlet_Size(object))

Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [33]:
# fill null values for object columns
df['Outlet_Size'] = df['Outlet_Size'].fillna('MISSING')
# check number of NaN
df['Outlet_Size'].isna().sum()

0

In [35]:
# fill null values for numeric column
df['Item_Weight'] = df['Item_Weight'].fillna(-1)
# check number of NaN
df['Item_Weight'].isna().sum()

0

In [36]:
df.isna().sum()
# no null values

Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

In [38]:
# check consistency
# object columns
object_col = df.select_dtypes('object').columns
for col in object_col:
  print(f'Value Count for {col}\n{df[col].value_counts()}\n ')
# from details below you can find that Item_Fat_content has inconsistent values

Value Count for Item_Fat_Content
Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64
 
Value Count for Item_Type
Fruits and Vegetables    1232
Snack Foods              1200
Household                 910
Frozen Foods              856
Dairy                     682
Canned                    649
Baking Goods              648
Health and Hygiene        520
Soft Drinks               445
Meat                      425
Breads                    251
Hard Drinks               214
Others                    169
Starchy Foods             148
Breakfast                 110
Seafood                    64
Name: Item_Type, dtype: int64
 
Value Count for Outlet_Identifier
OUT027    935
OUT013    932
OUT049    930
OUT046    930
OUT035    930
OUT045    929
OUT018    928
OUT017    926
OUT010    555
OUT019    528
Name: Outlet_Identifier, dtype: int64
 
Value Count for Outlet_Size
Medium     2793
MISSING    2410
Small      2388
High        932
Name

In [41]:
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace(['LF','low fat'], 'Low Fat')
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('reg','Regular')
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [45]:
numeric_col = df.select_dtypes('number').columns
df.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8523.000000      8523.000000  8523.000000                8523.000000   
mean     10.478936         0.066132   140.992782                1997.831867   
std       6.720742         0.051598    62.275067                   8.371760   
min      -1.000000         0.000000    31.290000                1985.000000   
25%       6.650000         0.026989    93.826500                1987.000000   
50%      11.000000         0.053931   143.012800                1999.000000   
75%      16.000000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800

## Exploratory Data Analysis

## Feature Inspection